src/main/java/org/subethamail/smtp/client/SMTPClient.java

```diff
@@ -162,6 +186,20 @@ public void connect(String host, int port) throws IOException {
         this.socket.bind(this.bindpoint.orElse(null));
         this.socket.setSoTimeout(REPLY_TIMEOUT);
         this.socket.connect(new InetSocketAddress(host, port), CONNECT_TIMEOUT);
+        updateSocket(this.socket);
+        connected = true;
+    }
+
+    /**
+     * Replaces the existing socket with a new Socket, updating the readers, writers and streams derived
+     * from the socket.  Used on the initial connect() to set up IO, and on upgrading a connection to TLS,
+     * where the new Socket is an SSLSocket.
+     *
+     * @param socket
+     * @throws IOException
+     */
+    protected void updateSocket(Socket socket) throws IOException {
```

Comment:
whenever you call this you pass it `this.socket`. For this reason I'd like to see the `socket` parameter removed from this call and also obviously the assignment.